python/ray/util/client/logsclient.py

```diff
@@ -18,47 +21,68 @@
 
 
 class LogstreamClient:
-    def __init__(self, channel: "grpc._channel.Channel", metadata: list):
+    def __init__(self, client_worker, metadata: list):
         """Initializes a thread-safe log stream over a Ray Client gRPC channel.
 
         Args:
+            client_worker: The Ray Client worker that manages this client
             channel: connected gRPC channel
             metadata: metadata to pass to gRPC requests
         """
-        self.channel = channel
+        self.client_worker = client_worker
         self._metadata = metadata
         self.request_queue = queue.Queue()
         self.log_thread = self._start_logthread()
         self.log_thread.start()
+        self.last_req = None
 
     def _start_logthread(self) -> threading.Thread:
         return threading.Thread(target=self._log_main, args=(), daemon=True)
 
     def _log_main(self) -> None:
-        stub = ray_client_pb2_grpc.RayletLogStreamerStub(self.channel)
-        log_stream = stub.Logstream(
-            iter(self.request_queue.get, None), metadata=self._metadata)
-        try:
-            for record in log_stream:
-                if record.level < 0:
-                    self.stdstream(level=record.level, msg=record.msg)
-                self.log(level=record.level, msg=record.msg)
-        except grpc.RpcError as e:
-            if e.code() == grpc.StatusCode.CANCELLED:
-                # Graceful shutdown. We've cancelled our own connection.
-                logger.info("Cancelling logs channel")
-            elif e.code() in (grpc.StatusCode.UNAVAILABLE,
-                              grpc.StatusCode.RESOURCE_EXHAUSTED):
-                # TODO(barakmich): The server may have
-                # dropped. In theory, we can retry, as per
-                # https://grpc.github.io/grpc/core/md_doc_statuscodes.html but
-                # in practice we may need to think about the correct semantics
-                # here.
-                logger.info("Server disconnected from logs channel")
-            else:
-                # Some other, unhandled, gRPC error
-                logger.exception(
-                    f"Got Error from logger channel -- shutting down: {e}")
+        reconnecting = False
```

Comment:
What is the main difference between _log_main() and _data_main() now?

Reply:
Mostly in how reconnection is handled. _data_main() is responsible for reconnecting the channel, while _log_main will just continue retrying until the channel is reconnected.